tmt/utils.py

```diff
@@ -3006,3 +3019,259 @@ def wait(
+            # Skip, needs fixes to become compatible
+            if klass is Common:
+                return
+
+            for name, value in klass.__dict__.get('__annotations__', {}).items():
+                # Skip special fields that are not keys.
+                if name == 'KEYS_SHOW_ORDER':
+                    continue
+
+                yield (name, value)
+
+        for base in cls.__mro__:
+            yield from _iter_class_annotations(base)
+
+        yield from _iter_class_annotations(cls)
+
+    @classmethod
+    def keys(cls) -> Generator[str, None, None]:
+        """
+        Iterate over key names.
+        Keys are yielded in the order: keys declared by parent classes first, then
+        keys declared by the class itself, all following the order in which keys
+        were defined in their respective classes.
+        Yields:
+            key names.
+        """
+
+        for keyname, _ in cls._iter_key_annotations():
+            yield keyname
+
+    def items(self) -> Generator[Tuple[str, Any], None, None]:
+        """
+        Iterate over keys and their values.
+        Keys are yielded in the order: keys declared by parent classes first, then
+        keys declared by the class itself, all following the order in which keys
+        were defined in their respective classes.
+        Yields:
+            pairs of key name and its value.
+        """
+
+        for keyname in self.keys():
+            yield (keyname, getattr(self, keyname))
+
+    # TODO: exists for backward compatibility for the transition period. Once full
+    # type annotations land, there should be no need for extra _keys attribute.
+    @classmethod
+    def _keys(cls) -> List[str]:
+        """ Return a list of names of object's keys. """
+
+        return list(cls.keys())
+
+    def _load_keys(
+            self,
+            key_source: Dict[str, Any],
+            key_source_name: str,
+            logger: Common) -> None:
+        """ Extract values for class-level attributes, and verify they match declared types. """
+
+        LOG_SHIFT, LOG_LEVEL = 2, 4
+
+        debug_intro = functools.partial(logger.debug, shift=LOG_SHIFT - 1, level=LOG_LEVEL)
+        debug = functools.partial(logger.debug, shift=LOG_SHIFT, level=LOG_LEVEL)
+
+        debug_intro('key source')
+        for k, v in key_source.items():
+            debug(f'{k}: {v} ({type(v)})')
+
+        debug('')
+
+        for keyname, keytype in self._iter_key_annotations():
+            key_address = f'{key_source_name}:{keyname}'
+            source_keyname = key_to_option(keyname)
+
+            # Do not indent this particular entry like the rest, so it could serve
+            # as a "header" for a single key processing.
+            debug_intro('key', key_address)
+            debug('field', source_keyname)
+
+            debug('desired type', str(keytype))
+
+            value: Any = None
+
+            debug('dict', self.__dict__)
```

Comment:
What is the point of debugging the whole dict here?

    key: /example:component
        field: component
        desired type: typing.List[str]
        dict: {'path': '/example', '_workdir': None, 'contact': [], 'component': [], 'test': './test.sh', 'framework': 'beakerlib', 'manual': False, 'require': [], 'recommend': [], 'environment': {}, 'duration': '5m', 'result': 'respect', 'summary': 'Concise summary describing what the test does', 'description': None, 'enabled': True, 'order': 50, 'link': <tmt.base.Link object at 0x7f616e0b2a70>, 'id': None, 'tag': [], 'tier': None, 'adjust': [None]}
        detected mutable default
        default value: []

In this way it is repeated for each normalized key again and again.

Reply:
Yeah, it's probably not that important, it just improves the overall picture of how's normalization progressing. It's repeated, yes, but it should be different (almost) every time it's printed because the previous key probably added a key to `self.__dict__`.

I'd like to keep it, but it can surely be buried on a deeper level, how about that?

Reply:
I see, thanks for clarifying. Let's move it one level deeper.

Reply:
Done.